### DISCUSSION TITLE
[RFC] 063 - Agent Provider

### DISCUSSION THREAD

**Discussion by arvinxx** [lobehub/lobehub#4444] [RFCs]
## 背景

model provider 的架构设计大获成功，在社区共同建设下，我们支持了 30+ 的 model providers。但单纯的 model 接入对于很多现有的平台来说还是不够的。例如用户可以 dify 上完成整个工作流的搭建和 agent 的配置，同时提供出来了完整的 api 接口。

![image](https://github.com/user-attachments/assets/19b44355-6eef-4554-adb7-6a07db914eef)

同时 OpenAI assistant 系列接口也提供了完整的 agent 能力，诸多 provider 也提供了对标 OpenAI assistant api 的接入实现

于是我们开始思考，是否有可能将 model provider 的设计思路延续到 agent 上。让 LobeChat 除了切换 model，也支持使用不同来源的 agent 。以及在后续的功能中将这个特性发挥出来。

在这样的设计中，我们可以将 Dify/OpenAI/Coze 等平台理解为 Agent 的生产端，而 LobeChat 就可以作为 Agent 的统一消费端，进而提供最佳的 agent 对话体验。

## 设计思路

- 交互上做到允许用户添加 dify/coze/openai assistant 上创建了的 agent。
- 像 dify 的执行也是有 CoT 的，如果能对接好 dify 的 api 接口，这样对 CoT 的理解就足够深入

agent 和 model 最不一样的地方就是 agent 的方案是状态持久化的，无论是 OpenAI 的 assistant api 还是 dify/coze 的助手，数据都是存在这些平台方上。LobeChat 的助手也是有状态的，所以有状态的这些接口其实就不像model那样那么容易扩展。但能力会更加强，比如支持上传文件，可以保留历史消息记录等等。

## 实现路径

- 首先研究 Dify agent provider ，#4412 ，实现 Dify 的 agent 接入与使用；
- 在此基础上添加 OpenAI 的 assistant API 
- 当有了 LobeChat / OpenAI assistant / Dify 时，我们就可以开始抽象接入一个 Agent Provider 的模式

**Comment by SoulGolf** [comment]
很期待dify的原生接入,我现在就为了接入dify,得先dify2openai,再接入oneapi,oneapi再接入LobeChat,对历史上下文和文件支持也不好

**Reply by vip2193** [reply]
我和你的思路一样
